modules/asg/variables.tf

```diff
@@ -96,14 +96,14 @@ variable "root_volume_size" {
 variable "additional_block_devices" {
   default     = []
   description = "Additional EBS volumes to use as instance block devices"
-  type        = list(object({device_name = string, volume_type = string, volume_size = string }))
+  type        = list(object({ device_name = string, volume_type = string, volume_size = string, encrypted = bool }))
```

Comment:
There is a problem with optional attributes in HCL - https://github.com/hashicorp/terraform/issues/19898
If we specify it here then user will need to supply some value for `encrypted` (even if that value is `null`). As HCL is dynamically typed there is another way around: not specify optional attributes while still using them.

Reply:
That is a good point. But, if I omit it here, hence user does not give a value for "encrypted". Then when I use it anyway, what would the value be?

Reply:
@Magicloud probably before I was doing something incorrectly but I don't have that code anymore. The current test:
```
variable "obj" {
  type = object({name = string, key = string})
}
resource "null_resource" "o" {
  provisioner "local-exec" {
    command = "echo ${var.obj.unspecified}"
  }
}
```
gives me an error:
```
Error: Unsupported attribute

  on main.tf line 26, in resource "null_resource" "o":
  26:     command = "echo ${var.obj.unspecified}"

This object does not have an attribute named "unspecified".
```
But replacing `var.obj.unspecified` with `lookup(var.obj, "unspecified", "default")` gives a better result:
```
$ terraform apply
var.obj
  Enter a value: {"name":"boo","key": 42}

null_resource.u[1]: Refreshing state... [id=2952313068434605786]
null_resource.o: Refreshing state... [id=5465750222682703737]
null_resource.u[0]: Refreshing state... [id=4296688519603364011]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # null_resource.o is tainted, so must be replaced
-/+ resource "null_resource" "o" {
      ~ id = "5465750222682703737" -> (known after apply)
    }

Plan: 1 to add, 0 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

null_resource.o: Destroying... [id=5465750222682703737]
null_resource.o: Destruction complete after 0s
null_resource.o: Creating...
null_resource.o: Provisioning with 'local-exec'...
null_resource.o (local-exec): Executing: ["/bin/sh" "-c" "echo default"]
null_resource.o (local-exec): default
null_resource.o: Creation complete after 0s [id=1869318344897596251]
```

Reply:
Great. Let me try some.

Reply:
Not sure if I did something wrong. The default value did not work for me.
module: `encrypted   = lookup(ebs_block_device.value, "encrypted", false)`, caller: `additional_block_devices = [{device_name = "/dev/sdl", volume_type = "gp2", volume_size = 15, encrypted = true}]`. The generated plan and applied result was still not encrypted.

Reply:
yes, it looks like TF keeps only attributes you specify, that's very unfortunate :(
So then we could only specify everything explicitly (using `null` when we can't)